compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1585,11 +1585,12 @@ object Types {
 
     protected def asMemberOf(prefix: Type)(implicit ctx: Context): Denotation =
       if (name.isShadowedName) prefix.nonPrivateMember(name.revertShadowed)
-      else {
-        val d = lastDenotation
-        // Never go from a non-private denotation to a private one
-        if (d == null || d.symbol.is(Private)) prefix.member(name)
-        else prefix.nonPrivateMember(name)
+      else lastDenotation match {
+        case d: SymDenotation if !d.is(Private) =>
+          // We shouldn't go from a non-private denotation to a private one
+          prefix.nonPrivateMember(name)
+        case _ =>
+          prefix.member(name)
```

Comment:
On second thought, I think we should revert to the previous scheme where `allowPrivate` was passed as a boolean to `asMemberOf`. There are 3 calls to `asMemberOf`. The ones in `recomputeDenot` and `newLikeThis` can get the parameter from the symbol. The third one, in `loadDenot` should pass `false` for
`allowPrivate`. After all, it could be that a definition was public in run 1, but private in run 2. We need to be able to pick up the new definition in `loadDenot`.